ledger-service/fetch-contracts/src/main/scala/fetchcontracts/domain.scala

```diff
@@ -96,24 +99,45 @@ package domain {
       a.key.fold(false)(_ == k)
 
     def fromLedgerApi(
-        gacr: lav1.active_contracts_service.GetActiveContractsResponse
+        resolvedQuery: domain.ResolvedQuery,
+        gacr: lav1.active_contracts_service.GetActiveContractsResponse,
     ): Error \/ List[ActiveContract[lav1.value.Value]] = {
-      gacr.activeContracts.toList.traverse(fromLedgerApi(_))
+      gacr.activeContracts.toList.traverse(fromLedgerApi(resolvedQuery, _))
     }
 
-    def fromLedgerApi(in: lav1.event.CreatedEvent): Error \/ ActiveContract[lav1.value.Value] =
+    def fromLedgerApi(
+        resolvedQuery: domain.ResolvedQuery,
+        in: lav1.event.CreatedEvent,
+    ): Error \/ ActiveContract[lav1.value.Value] = {
+
+      def getIdAndPayload: (Error \/ ContractTypeId.Resolved, Error \/ lav1.value.Record) =
+        resolvedQuery match {
+          case ResolvedQuery.ByInterfaceId(interfaceId) =>
+            import util.IdentifierConverters.apiIdentifier
+            val id = apiIdentifier(interfaceId)
+            val payload = in.interfaceViews
+              .find(_.interfaceId.exists(_ == id))
+              .flatMap(_.viewValue) required "interfaceView"
+            (\/-(ContractTypeId.Interface fromLedgerApi id), payload)
+          case _ =>
+            val id = in.templateId.required("templateId").map(ContractTypeId.Template.fromLedgerApi)
+            (id, in.createArguments required "createArguments")
+        }
+
+      val (getId, getPayload) = getIdAndPayload
       for {
-        templateId <- in.templateId required "templateId"
-        payload <- in.createArguments required "createArguments"
+        id <- getId
+        payload <- getPayload
       } yield ActiveContract(
         contractId = ContractId(in.contractId),
-        templateId = ContractTypeId.Template fromLedgerApi templateId,
+        templateId = id,
         key = in.contractKey,
```

Comment:
It occurs to me while building #14920 that returning `contractKey` isn't in the spirit of the API we're offering. Just as we replace the `templateId` with interface ID in the by-interface case, we should also put `None` here, unless and until we have interface keys, in which case that should go here, rather than the template-associated key.

It's mostly harmless that it's wrong, but we shouldn't support it, and putting `None` here in the interface case is the best way to ensure that we don't.

Reply:
@S11001001 fixed here: https://github.com/digital-asset/daml/pull/14917/commits/7ee8038a013b968ea144ca8d4d659d1fe664ce22